### PR TITLE
fix: upload-artifact and download-artifact v4

### LINKED ISSUE #3068: [feature] Use upload-artifact v4
we may need a compatibility option as argument to avoid breaking folks who currently use v3

### PR DESCRIPTION
**Pull request by ramonpetgrave64**
# Summary

- Fixes #3068 to use upload-artifact and download-artifact v4
- following up in https://github.com/slsa-framework/example-package/pull/336

## Testing Process

This change is tested with our existing PR Check workflows that use both directly and indirectly call upload-artifact and download-artifact. 
 - One test for `secure-upload-folder` should fail in this PR because it will use `secure-upload-artifact@main`. There's no workaround to dynamically use the PR's ref instead of `@main`, but after merging this PR, the test should start passing.

## Checklist

- [x] Review the contributing [guidelines](./../CONTRIBUTING.md)
- [x] Add a reference to related issues in the PR description.
- [x] Update documentation if applicable.
- [x] Add unit tests if applicable.
- [x] Add changes to the [CHANGELOG](./../CHANGELOG.md) if applicable.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -99,6 +100,10 @@ duplication."
 
 ## Unreleased
 
+### Unreleased: Breaking Change: upload-artifact and download-artifact
+
+- Our workflows now use the new `@v4`s of `actions/upload-artifact` and `actions/download-artifact`, which are incompatiblle with the prior `@v3`. See Our docs on the [generic generator](./internal/builders/generic/README.md#compatibility-with-actionsdownload-artifact).
```
Comment (laurentsimon) [inline CHANGELOG.md:105]: nit: Add a sentence (or a section) on how to upgrade from v2 to v3.  know it looks obvious, but let's call it out explicitly. (We can clean up the "how to upgrade" into. dedicated section when we do the release if needed) Wdut?

File: .github/workflows/pre-submit.actions.yml [hunk 2]
```diff
@@ -28,12 +28,18 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+        with:
+          repository: ${{ github.event_name == 'pull_request' && github.event.pull_request.head.repo.full_name || github.repository }}
```
Comment (laurentsimon) [inline .github/workflows/pre-submit.actions.yml:32]: on pull_request events, the checkout Action knows to clone the fork's repo at the right hash, so we don't need to tell it explicitly. Right? Did you run into problems?

**Review by laurentsimon — COMMENTED** [review]
LGTM. Had a comment on the workflow update

**Review by laurentsimon — COMMENTED** [review]
Any reason you ad to update the package-lock.json? That should not be necessary. Maybe a leftover from prior experiments :)?

**Comment by ramonpetgrave64** [comment]
> Any reason you ad to update the package-lock.json? That should not be necessary. Maybe a leftover from prior experiments :)?

Just a package that had a vulnerability. I bumped the version with `npm audit fix`.

File: .github/workflows/pre-submit.actions.yml (hunk 2 above)
Comment (ramonpetgrave64) [reply .github/workflows/pre-submit.actions.yml:32]: I recall running into some problems. But I've done some more proper testing and you're right. It does cehckout the PR's code by default. I'll remove these changes.

**Comment by laurentsimon** [comment]
please check the linter warning

**Comment by ramonpetgrave64** [comment]
@laurentsimon I removed the npm change and fixed the makrdownlint errors

**Review by laurentsimon — APPROVED** [review]

**Comment by laurentsimon** [comment]
secure-upload-folder seems to be failing. can you take a look?

**Comment by ramonpetgrave64** [comment]
Yes, it should fail because

> * One test for `secure-upload-folder` should fail in this PR because it will use `secure-upload-artifact@main`. There's no workaround to dynamically use the PR's ref instead of `@main`, but after merging this PR, the test should start passing

**Comment by laurentsimon** [comment]
> * secure-upload-folder

Got it. I created https://github.com/slsa-framework/slsa-github-generator/issues/3320 for tracking. Should be easy to fix.

I'll disable the check temporarily to merge

**Comment by laurentsimon** [comment]
Any idea why `pre-submit e2e generic default / build / final (pull_request)` is failing?

**Pull request merged**